source/extensions/dynamic_modules/dynamic_modules.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace DynamicModules {
+
+/**
+ * A class for loading and managing dynamic modules. This corresponds to a single dlopen handle.
+ * When the DynamicModule object is destroyed, the dlopen handle is closed.
+ *
+ * This class is supposed to be initialized once in the main thread and can be shared with other
+ * threads.
+ */
+class DynamicModule {
```

Comment:
note: this should be utilized in contrib/golang, and relatively easy, but  then the changes in this patch end up across many places, so I chose not to do that right now. - I could take a shot after this PR.

Reply:
Yes please do, thanks.